Add ability to log when max_partitions_per_insert_block is reached

seandhaynes opened the pull request:
If you wish to reduce the value of  `max_partitions_per_insert_block` down from it's default of 100, 
it can difficult to do without knowing which inserts/clients would be impacted.

This commit adds the ability to log when the above limit is reached rather than throw an exception,
so that you can gauge the impact on users when dropping this value down.
Also incrementing the RejectedInserts ProfileEvent when this is encountered so rejections here are accounted for.


### Changelog category (leave one):
- Improvement



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to log when max_partitions_per_insert_block is reached
...

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #50948 from seandhaynes/ClickHouse:master -->
This is an automated comment for commit 2d316df5080236eed018032f340eb75418ee8958 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50948/2d316df5080236eed018032f340eb75418ee8958/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

nickitat commented on src/Storages/MergeTree/MergeTreeDataWriter.cpp:
```diff
@@ -75,16 +76,18 @@ void buildScatterSelector(
         if (inserted)
         {
             if (max_parts && partitions_count >= max_parts)
+
```
these lines should be wrapped into `{}`

seandhaynes commented on src/Storages/MergeTree/MergeTreeDataWriter.cpp:
```diff
@@ -75,16 +76,18 @@ void buildScatterSelector(
         if (inserted)
         {
             if (max_parts && partitions_count >= max_parts)
```
> we always increment it by one as far as I see, so shouldn't `partitions_count` always be equal to `max_parts` in case of this error?

Thanks @nickitat , you're correct.  I misread the iteration above. I will update this as well as fix the bug above.

seandhaynes committed:
Add ability to log when max_partitions_per_insert_block is reached

There's plenty of reasons to drop max_partitions_per_insert_block
down from it's default of 100, with the main one being even touching
remotely close to this number of partitions will create a lot of
merge activity that can de-stabilise a cluster.

This commit adds the ability to log when the above limit is reached
rather than throw an exception, so that you can gauge the impact on
users when dropping this value down.

seandhaynes commented:
Rebased after a conflict with: https://github.com/ClickHouse/ClickHouse/commit/e58b3cfd3246681092908dc6e38d0b9a130b5df7

seandhaynes commented:
@nickitat Hey, sorry to bother you again, but it looks like I only have a single failed build that looks unrelated: https://github.com/ClickHouse/ClickHouse/actions/runs/5714895086/job/15488863624?pr=50948
```
2023-07-31 19:49:30 exceptions.ProgramError: request doesn't have a result set, it has the exception, exception: ('HY000', '[HY000] HTTP status code: 500\nReceived error:\nCode: 373. DB::Exception: Session clickhouse_odbc_12551460960992802180 is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.8.1.250)\n\n (1) (SQLExecDirectW)')
+ cat /test_output/statements-test/check_status.tsv
cat: /test_output/statements-test/check_status.tsv: No such file or directory
+ cat /test_output/statements-test/test_results.tsv
cat: /test_output/statements-test/test_results.tsv: No such file or directory
+ tar -zcvf statements-check.tar.gz statements-test
+ clickhouse-client -q 'system flush logs'
```

Did you want us to re-run this, or can we merge?

seandhaynes commented:
🎉  Thanks very much for re-triggering, @nickitat .

seandhaynes commented:
@nickitat  thanks again. Now all builds are successful, are you happy merging?

nickitat commented on src/Storages/MergeTree/MergeTreeDataWriter.cpp:
```diff
@@ -102,6 +109,18 @@ void buildScatterSelector(
+        const auto & client_info = context->getClientInfo();
+        Poco::Logger * log = &Poco::Logger::get("MergeTreeDataWriter");
+
+        LOG_WARNING(log, "INSERT query from initial_user {} (query ID: {}) inserted a block "
```
it is not necessary to log client info - it is present in `query_log`

nickitat commented on docs/ru/operations/settings/query-complexity.md:
```diff
@@ -311,9 +311,18 @@ FORMAT Null;
 **Подробности**
 
-При вставке данных, ClickHouse вычисляет количество партиций во вставленном блоке. Если число партиций больше, чем `max_partitions_per_insert_block`, ClickHouse генерирует исключение со следующим текстом:
+При вставке данных ClickHouse проверяет количество разделов в вставляемом блоке. Если количество разделов превышает число `max_partitions_per_insert_block`, ClickHouse либо регистрирует, либо выбрасывает исключение на основе `throw_on_max_partitions_per_insert_block`.  Исключения имеют следующий текст:
```
```suggestion
При вставке данных ClickHouse проверяет количество партиций во вставляемом блоке. Если количество разделов превышает число `max_partitions_per_insert_block`, ClickHouse либо логирует предупреждение, либо выбрасывает исключение в зависимости от значения `throw_on_max_partitions_per_insert_block`.  Исключения имеют следующий текст:
```

nickitat commented on docs/ru/operations/settings/query-complexity.md:
```diff
@@ -311,9 +311,18 @@ FORMAT Null;
+
+## throw_on_max_partitions_per_insert_block {#settings-throw_on_max_partition_per_insert_block}
+
+Pozvolyayet vam kontrolirovat' povedeniye pri dostizhenii `max_partitions_per_insert_block`
```
```suggestion
Позволяет контролировать поведение при достижении `max_partitions_per_insert_block`
```

nickitat commented on docs/ru/operations/settings/query-complexity.md:
```diff
@@ -311,9 +311,18 @@ FORMAT Null;
+
+Pozvolyayet vam kontrolirovat' povedeniye pri dostizhenii `max_partitions_per_insert_block`
+
+- `true`  - Когда вставляемый блок достигает max_partitions_per_insert_block, возникает исключение.
```
```suggestion
- `true`  - Когда вставляемый блок достигает `max_partitions_per_insert_block`, возникает исключение.
```

seandhaynes committed:
Update docs/ru/operations/settings/query-complexity.md

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

seandhaynes committed:
Update docs/ru/operations/settings/query-complexity.md

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

seandhaynes committed:
Update docs/ru/operations/settings/query-complexity.md

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

nickitat commented:
`Stress test (ubsan)` - should be fixed by #53111

seandhaynes commented:
Thanks so much for your help with this and merging it, @nickitat
